### PR TITLE
[NFC][RemoveDIs] Switch constant-hoisting to insert with iterators

### PR DESCRIPTION
**Pull request by jmorse**
Seeing how constant-hoisting also happens to store an insertion-point in a DenseMap, this means we have to install DenseMapInfo for hashing BasicBlock::iterators and comparing them. I'm not really sure where to put the DenseMapInfo declarations as BasicBlock.h seems most logical, but that then means including DenseMap.h into pretty much all of LLVM. I've sent this up to the compile time tracker to see whether there's a major cost from this.

### CODE REVIEW

**Review by OCHyams — APPROVED** [review]
> I've sent this up to the compile time tracker to see whether there's a major cost from this.

LGTM, pending those results (and a clang-format)

**Comment by SLTozer** [comment]
> LGTM, pending those results (and a clang-format)

http://llvm-compile-time-tracker.com/compare.php?from=5b5c21d772d20320fd876edddfc204cca93fae0f&to=aaaf4583cca106efe86f23c7d6f4eb462624288c&stat=instructions:u

tl;dr no noticeable impact!

File: llvm/include/llvm/IR/BasicBlock.h [hunk 1]
```diff
@@ -774,6 +774,32 @@ BasicBlock::iterator skipDebugIntrinsics(BasicBlock::iterator It);
...
+  static inline BasicBlock::iterator getTombstoneKey() {
+    BasicBlock::iterator It(nullptr);
+    It.setHeadBit(true);
+    return It;
+  }
+
+  static unsigned getHashValue(const BasicBlock::iterator &It) {
+    return DenseMapInfo<void *>::getHashValue(
+               reinterpret_cast<void *>(It.getNodePtr())) ^
+           It.getHeadBit();
```
Comment (ScottTodd) [inline llvm/include/llvm/IR/BasicBlock.h:794]: This triggers warnings on MSVC: `llvm-project\llvm\include\llvm/IR/BasicBlock.h(794): warning C4805: '^': unsafe mix of type 'unsigned int' and type 'bool' in operation`

Comment (ScottTodd) [reply llvm/include/llvm/IR/BasicBlock.h:794]: Ah, there are PRs for that already.
